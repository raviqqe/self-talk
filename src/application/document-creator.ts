import UUID from "pure-uuid";
import { formatDocument, validateDocument } from "../domain/document";
import { formatErrorMessage } from "../domain/error";
import { IDocumentPresenter } from "./document-presenter";
import { IDocumentRepository } from "./document-repository";
import { IMessagePresenter } from "./message-presenter";

export class DocumentCreator {
  constructor(
    private readonly documentRepository: IDocumentRepository,
    private readonly documentPresenter: IDocumentPresenter,
    private readonly messagePresenter: IMessagePresenter
  ) {}

  public async create(text: string): Promise<void> {
    const document = formatDocument({ id: new UUID(4).format(), text });

    try {
      validateDocument(document);
    } catch (error) {
      this.messagePresenter.present(formatErrorMessage(error as Error));
      return;
    }

    this.documentPresenter.presentNewDocument(document);
    await this.documentRepository.create(formatDocument(document));
  }
}
