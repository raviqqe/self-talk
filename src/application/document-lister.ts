import { IDocument } from "../domain/document";
import { IDocumentPresenter } from "./document-presenter";
import { IDocumentRepository } from "./document-repository";

const defaultLimit = 20;

export class DocumentLister {
  private iterator: AsyncIterator<IDocument[], void> | null = null;

  constructor(
    private readonly documentRepository: IDocumentRepository,
    private readonly documentPresenter: IDocumentPresenter
  ) {}

  public async list(): Promise<void> {
    this.iterator = this.documentRepository.list(defaultLimit);
    this.documentPresenter.presentDocuments(
      (await this.iterator.next()).value || []
    );
  }

  public async listMore(): Promise<void> {
    if (!this.iterator) {
      throw new Error("iterator not initialized");
    }

    const result = await this.iterator.next();

    if (result.done) {
      return;
    }

    this.documentPresenter.presentMoreDocuments(result.value);
  }
}
