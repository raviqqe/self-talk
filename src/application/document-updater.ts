import {
  IDocument,
  formatDocument,
  validateDocument,
} from "../domain/document";
import { formatErrorMessage } from "../domain/error";
import { DocumentDeleter } from "./document-deleter";
import { IDocumentRepository } from "./document-repository";
import { IDocumentPresenter } from "./document-presenter";
import { IMessagePresenter } from "./message-presenter";

export class DocumentUpdater {
  constructor(
    private readonly documentDeleter: DocumentDeleter,
    private readonly documentRepository: IDocumentRepository,
    private readonly documentPresenter: IDocumentPresenter,
    private readonly messagePresenter: IMessagePresenter
  ) {}

  public async update(document: IDocument): Promise<void> {
    document = formatDocument(document);

    if (!document.text) {
      await this.documentDeleter.delete(document.id);
      return;
    }

    try {
      validateDocument(document);
    } catch (error) {
      this.messagePresenter.present(formatErrorMessage(error));
      return;
    }

    this.documentPresenter.presentUpdatedDocument(document);
    await this.documentRepository.update(document);
  }
}
