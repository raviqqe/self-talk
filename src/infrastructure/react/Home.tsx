import React from "react";
import styled from "styled-components";
import { CreateDocument } from "./CreateDocument";
import { Documents, IProps as IDocumentsProps } from "./Documents";
import { SignOut } from "./SignOut";
import { darkYellow } from "./style/colors";

const Container = styled.div`
  display: flex;
  flex-direction: column;
  height: 100vh;
  width: 85ex;
  max-width: 100%;
  margin: auto;

  > :first-child {
    flex: 1;
  }
`;

const CreateDocumentContainer = styled.div`
  position: relative;

  * {
    z-index: 1;
  }
`;

const CreateDocumentBackground = styled.div`
  background: ${darkYellow};
  position: absolute;
  top: 0;
  left: -100vw;
  width: 200vw;
  height: 100%;
  box-shadow: 0rem 0rem 0.5rem rgba(0, 0, 0, 0.3);
  z-index: 0;
`;

const SignOutContainer = styled.div`
  position: fixed;
  top: 0.5em;
  right: 0.5em;
`;

export interface IProps extends IDocumentsProps {
  createDocument: (text: string) => Promise<void>;
  signOut: () => void;
}

export const Home = ({
  createDocument,
  documents,
  insertFiles,
  listDocuments,
  listMoreDocuments,
  signOut,
  updateDocument,
}: IProps) => (
  <Container>
    <Documents
      documents={documents}
      insertFiles={insertFiles}
      listDocuments={listDocuments}
      listMoreDocuments={listMoreDocuments}
      updateDocument={updateDocument}
    />
    <CreateDocumentContainer>
      <CreateDocument
        createDocument={createDocument}
        insertFiles={insertFiles}
      />
      <CreateDocumentBackground />
    </CreateDocumentContainer>
    <SignOutContainer>
      <SignOut signOut={signOut} />
    </SignOutContainer>
  </Container>
);
