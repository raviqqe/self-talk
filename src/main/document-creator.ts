import { DocumentCreator } from "../application/document-creator.js";
import { documentPresenter } from "./document-presenter.js";
import { documentRepository } from "./document-repository.js";
import { messagePresenter } from "./message-presenter.js";

export const documentCreator = new DocumentCreator(
  documentRepository,
  documentPresenter,
  messagePresenter,
);
