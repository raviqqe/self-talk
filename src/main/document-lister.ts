import { DocumentLister } from "../application/document-lister.js";
import { documentPresenter } from "./document-presenter.js";
import { documentRepository } from "./document-repository.js";

export const documentLister = new DocumentLister(
  documentRepository,
  documentPresenter,
);
