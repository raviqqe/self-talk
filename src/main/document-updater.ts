import { DocumentUpdater } from "../application/document-updater.js";
import { documentDeleter } from "./document-deleter.js";
import { documentPresenter } from "./document-presenter.js";
import { documentRepository } from "./document-repository.js";
import { messagePresenter } from "./message-presenter.js";

export const documentUpdater = new DocumentUpdater(
  documentDeleter,
  documentRepository,
  documentPresenter,
  messagePresenter,
);
