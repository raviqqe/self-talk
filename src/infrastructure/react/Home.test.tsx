import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { beforeEach, expect, it, vi } from "vitest";
import { documentCreator } from "../../main/document-creator.js";
import { documentLister } from "../../main/document-lister.js";
import { documentUpdater } from "../../main/document-updater.js";
import { Home } from "./Home.js";

beforeEach(() => {
  vi.spyOn(documentLister, "list").mockImplementation(async () => {});
  vi.spyOn(documentLister, "listMore").mockImplementation(async () => {});
});

it("renders", async () => {
  const result = await act(async () => render(<Home documents={[]} />));

  await waitFor(() => expect(screen.getByRole("textbox")).toBeTruthy());

  expect(result.container.firstChild).toMatchSnapshot();
});

it("creates a document", async () => {
  const createDocument = vi
    .spyOn(documentCreator, "create")
    .mockImplementation(async () => {});

  act(() => {
    render(<Home documents={[]} />);
  });

  await waitFor(() => expect(screen.getByRole("textbox")).toBeTruthy());

  act(() => {
    fireEvent.change(screen.getByRole("textbox"), {
      target: { value: "foo" },
    });

    fireEvent.click(screen.getByLabelText("Create"));
  });

  expect(createDocument.mock.calls).toHaveLength(1);
});

it("updates a document", async () => {
  const updateDocument = vi
    .spyOn(documentUpdater, "update")
    .mockImplementation(async () => {});

  const result = await act(async () =>
    render(<Home documents={[{ id: "", text: "" }]} />),
  );

  await waitFor(() => expect(result.getByLabelText("Edit")).toBeTruthy());

  act(() => {
    fireEvent.click(result.getByLabelText("Edit") as Element);

    fireEvent.change(result.container.querySelector("textarea")!, {
      target: { value: "foo" },
    });

    fireEvent.click(result.getByLabelText("Save") as Element);
  });

  expect(updateDocument.mock.calls).toHaveLength(1);
});
