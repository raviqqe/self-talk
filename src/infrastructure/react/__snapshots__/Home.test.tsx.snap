// Vitest Snapshot v1

exports[`renders 1`] = `
<div
  class="sc-ksBlkl ekagmZ"
>
  <div
    class="sc-ipEyDJ jJxCCy"
    id="documents-container"
  >
    <div
      class="infinite-scroll-component__outerdiv"
    >
      <div
        class="infinite-scroll-component sc-eDWCr jVClKC"
        style="height: auto; overflow: auto;"
      />
    </div>
  </div>
  <div
    class="sc-hBxehG dwmxpD"
  >
    <div
      class="sc-eDvSVe dSKcyM"
    >
      <textarea
        class="sc-dkrFOg bOiUUM sc-jSUZER JcOjq"
        placeholder="Write in Markdown ..."
      />
      <button
        aria-label="Create"
        class="sc-bcXHqe sc-gswNZR eTvdtE gWKlTj"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </button>
    </div>
    <div
      class="sc-fnGiBr dolgJS"
    />
  </div>
  <div
    class="sc-fEXmlR bXmvX"
  >
    <button
      class="sc-bcXHqe sc-gswNZR fmEPPj gWKlTj"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5a2 2 0 00-2 2v4h2V5h14v14H5v-4H3v4a2 2 0 002 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </button>
  </div>
</div>
`;
